module/zfs/dbuf.c

```diff
@@ -3842,6 +3873,10 @@ dbuf_sync_leaf(dbuf_dirty_record_t *dr, dmu_tx_t *tx)
 		    DN_MAX_BONUS_LEN(dn->dn_phys));
 		DB_DNODE_EXIT(db);
 
+#ifdef DEBUG
```

Comment:
You want `ZFS_DEBUG` here, and you'll need to wrap the new `dbuf_sync_leaf_verify_bonus_dnode()` with it as well to keep `gcc -Wall` happy for non-debug builds.